Fix connection leak caused by exceptions during rollback.

briancecker opened the pull request:
If an exception is thrown while attempting to rollback a transaction,
then the connection is never returned to the connection pool. This
slowly leaks connections until there are none left.

briancecker commented:
The heap dump below shows the state of a connection pool that has been completely allocated out. 
![pl_id_syncer_connections](https://user-images.githubusercontent.com/3677540/65074543-ef2c2b00-d949-11e9-8942-274f0c33d1b8.png)

If we further inspect these allocated objects, we can see that in our case, they were borrowed nearly two days ago, and the logs show that at exactly the time when they were last borrowed they threw an `Exception` during a transaction, and then failed to ever execute a DB query again. 
![Screen Shot 2019-09-17 at 10 16 23 AM](https://user-images.githubusercontent.com/3677540/65075258-7037f200-d94b-11e9-9b8f-7a05a4a72d90.png)

The epoch time for `lastBorrowTime` corresponds to September 15, 2019 11:23:37, exactly the time in our logs where we start to see consistent errors until restart. 
<img width="1453" alt="Screen Shot 2019-09-17 at 10 18 33 AM" src="https://user-images.githubusercontent.com/3677540/65075318-8a71d000-d94b-11e9-8571-1180dce23084.png">

joshcurtis commented on jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java:
```diff
@@ -96,7 +96,14 @@ DbMetrics getDbMetrics() {
       if (asTransaction) {
-        connection.rollback();
+        try {
+          connection.rollback();
```
Does it make sense to put this in a `try-finally` block? And remove this block below since the connection was returned?
``` java
if (asTransaction) {
  connection.rollback();
}
```

briancecker replied:
Do you mean remove the `catch (Throwable t)` block below and add a `finally` block? I think it makes sense to do that, but there's this big block comment above that explicitly mentions why we don't do that. I sort of disagree with that comment though, so I'm hoping @sherifnada can weigh in as he added that comment and logic.

hswaffield replied:
I think josh is saying we should always return the connection, regardless of how the rollback went, so it makes sense to just put that particular piece inside a `finally` block, once, and put the rollback inside a `try` block... that way we have `dbManager.returnConnection(connection);` written only once here.

I agree with that idea.

briancecker replied:
I think this should be resolved now with the recent changes.

sherifnada replied:
Hey guys, just catching up after coming back from a trip to China. It looks like this was merged already -- @briancecker did you feel that comment was inaccurate? or that it's the wrong way of fixing that problem?

hswaffield replied:
We believe it was the wrong way to fix the problem because it resulted in the `DBPoolManager` never letting go of stale connections, in certain situations, which killed a number of our apps as they could not get any new connections.

briancecker replied:
As Henry mentioned, we didn't so much feel it was inaccurate, but the way of solving the issue lead to another issue. To preserve the behavior you added, avoid committing half-finished transactions, and avoid leaking connections, I added a new method for invalidating a connection instead of returning it. This method (`DbPoolManager#invalidateConnection`) removes the connection from the pool's tracked objects and closes it without committing.

h-wang94 commented on jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java:
```diff
@@ -96,7 +96,14 @@ DbMetrics getDbMetrics() {
+        try {
+          connection.rollback();
+        } catch (Exception e2) {
+          // ignore. We can't actually do anything about this, and we need to return the connection
```
Is it worth logging that we were unable to rollback?

briancecker replied:
Oh, yeah that might be useful. I'll log it.

h-wang94 commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbTransactorImpl.java:
```diff
@@ -38,6 +44,43 @@ public void cleanup() throws Exception {
+      transactor.queryAsTransaction(db -> db.users().find(1));
+      fail();
+    } catch (Exception e) {
+      // This should fail
```
Probably need to fix up the wording here. It currently seems like its implying something should be failing inside this catch block instead of implying that this is the correct behavior.

briancecker replied:
Changed it. Hopefully it's more clear now.

h-wang94 commented on jack-test/test/java/com/rapleaf/jack/transaction/TestDbTransactorImpl.java:
```diff
@@ -38,6 +44,43 @@ public void cleanup() throws Exception {
+    Mockito.verify(dbPoolManager, Mockito.times(1)).returnConnection(Mockito.any(IDatabase1.class));
+
+    // This simply needs to not throw, demonstrating that the connection pool is not exhausted.
+    transactor.query(db -> db.users().find(1));
```
I hate a bajillion magic numbers. 

Can we perhaps assign `1` to some variable at the top of this test and then re-use it so its clear what the relationship of the values are?

briancecker replied:
Ah. Yeah, I can assign it to a variable and leave a comment. It basically has no relation. I just need to query for something. It doesn't matter that it happens to be the same id in both the queries.

hswaffield commented:
I didn't previously see the comments that Sherif left before... I look forward to hearing more of his thoughts here.

briancecker commented:
Hmm, I've come up with a slightly different solution here. Instead of always returning the connections, how about we expose a version of the `ObjectPool#invalidateObject` method in the `IDbManager` interface as something like `IDbManager#invalidateConnection`? In this way we can explicitly destroy the connection and remove it from the pool without having to rely on any validation checks in the pool to later remove the invalid connection and without accidentally committing any half-finished transactions (although I think that is unlikely).

briancecker commented:
I went ahead and introduced `IDbManager#invalidateConnection` and modified `TransactorImpl` to make use of it as I mentioned above. I think the code is quite a bit cleaner now, and it makes a lot more sense logically too! Please review again if you have time. I'd love to get this merged today :)

briancecker commented:
> Just to confirm, by just invalidating the `<DB extends IDb> connection` we won't accidentally leave connections open on the DB side, will we? 

Good question! Calling the `DbPoolManager#invalidateConnection` eventually calls `DbPoolFactory#destroyObject`, which is defined as:
```
    @Override
    public void destroyObject(PooledObject<DB> connection) throws IOException {
      connection.getObject().close();
    }
```

So it should close the connection. Of course that `.close()` call could fail for some reason, but then we're out of options, and we just have to wait until the connection expires on the server if it somehow ended up orphaned, but I don't actually know how that would happen.
